### PR TITLE
EL-4252 - Menu: close menu when focus leaves the trigger button

### PR DESCRIPTION
**Pull request by JonathanToddMF**
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licensed under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-4252

#### Description of Proposed Changes

1.  Added new event listener to menu, listening for Tab presses to close the menu. Listening for a blur event causes the menu to close as the focus was removed from the opening button when going through the menu.
2. Added tests to cover new functionality

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4252-menu-close-menu-when-lose-focus

#### Downstream Build(s)

* [caf/ux-aspects-micro-focus](https://jenkins.swinfra.net/job/SEPG/job/caf/job/caf~ux-aspects-micro-focus~EL-4252-menu-close-menu-when-lose-focus~CI/)

### CODE REVIEW

File: src/components/menu/menu-trigger/menu-trigger.directive.ts [hunk 1]
```diff
@@ -421,4 +429,22 @@ export class MenuTriggerDirective implements OnInit, OnDestroy {
         }
     }
 
+    /** Check whether the overlay or button has focus */
+    private checkFocus(): boolean {
```
Comment (donna-marijne) [inline src/components/menu/menu-trigger/menu-trigger.directive.ts:433]:
```suggestion
    private hasFocus(): boolean {
```

File: src/components/menu/menu-trigger/menu-trigger.directive.ts [hunk 2]
```diff
@@ -421,4 +429,22 @@ export class MenuTriggerDirective implements OnInit, OnDestroy {
...
+        return this._overlayRef.hostElement.contains(document.activeElement) ||
+            this._elementRef.nativeElement.contains(document.activeElement);
+    }
+
+    /** Close the menu if there is no element focused */
+    private closeOnBlur(): void {
+        if (this.menu.isMenuOpen) {
+            timer(this._debounceTime)
```
Comment (donna-marijne) [inline src/components/menu/menu-trigger/menu-trigger.directive.ts:441]: I think `setTimeout` might be more efficient.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Thanks for the change. Looks like there is an issue with submenus when accessing via the keyboard.

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4252-menu-close-menu-when-lose-focus/#/components/buttons#dropdowns

1. Tab to the menu button.
2. Press enter.
3. Press right arrow.

The menu should not collapse. Could you add test coverage for this case?

After blur, if you walk the ancestors of `document.activeElement` and you don't find the class `ux-menu` you can conclude that focus has exited the menu tree.

File: e2e/pages/app/menu/menu.testpage.component.ts [hunk 3]
```diff
@@ -8,7 +8,7 @@ import { BehaviorSubject } from 'rxjs';
 })
 export class MenuTestPageComponent {
 
-    placement$: BehaviorSubject<string> = new BehaviorSubject<string>('left');
+    placement$: BehaviorSubject<string> = new BehaviorSubject<string>('right');
```
Comment (donna-marijne) [inline e2e/pages/app/menu/menu.testpage.component.ts:11]: Not clear why this was changed.

File: e2e/tests/components/menu/menu.e2e-spec.ts [hunk 4]
```diff
@@ -30,16 +31,48 @@ describe('Menu', () => {
     it('should change menu placement to top after initialization', async () => {
         await page.placementTopBtn.click();
         await page.openMenuBtn.click();
+
+        await browser.sleep(500);
```
Comment (donna-marijne) [inline e2e/tests/components/menu/menu.e2e-spec.ts:35]: Presumably this should be removed.

File: e2e/tests/components/menu/menu.e2e-spec.ts [hunk 5]
```diff
@@ -30,16 +31,48 @@ describe('Menu', () => {
...
+        it('should not close the menu when entering the submenu', async () => {
+            // tab to menu and open it with the keyboard
+            await page.topFocusBtn.click();
+            await browser.actions().sendKeys(Key.TAB).perform();
+            await browser.actions().sendKeys(Key.SPACE).perform();
+            await browser.actions().sendKeys(Key.ARROW_RIGHT).perform();
+            await browser.actions().sendKeys(Key.ARROW_RIGHT).perform();
+
+            // check menu open
+            expect(await element(by.className('ux-menu')).isPresent()).toBe(true);
```
Comment (donna-marijne) [inline e2e/tests/components/menu/menu.e2e-spec.ts:74]: Should probably check that the submenu is present as well.

File: src/components/menu/menu-trigger/menu-trigger.directive.ts [hunk 6]
```diff
@@ -421,4 +429,28 @@ export class MenuTriggerDirective implements OnInit, OnDestroy {
         }
     }
 
+    /** Check whether the overlay or button has focus */
```
Comment (donna-marijne) [inline src/components/menu/menu-trigger/menu-trigger.directive.ts:432]: This now only checks the overlay, so the comment (and probably function name) needs updating.

File: src/components/menu/menu-trigger/menu-trigger.directive.ts [hunk 7]
```diff
@@ -421,4 +429,28 @@ export class MenuTriggerDirective implements OnInit, OnDestroy {
...
+        });
+
+        return check;
+    }
+
+    /** Close the menu if there is no element focused */
+    private closeOnBlur(): void {
+        if (this.menu.isMenuOpen) {
+            setTimeout(() => {
+                console.log(this.hasFocus());
```
Comment (donna-marijne) [inline src/components/menu/menu-trigger/menu-trigger.directive.ts:449]: Remove this.

File: e2e/pages/app/menu/menu.testpage.component.html [hunk 8]
```diff
@@ -15,30 +17,95 @@
...
 
-        <button type="button" uxMenuItem>
+        <button id="menu-item-4" type="button" uxMenuItem>
             <span class="dropdown-menu-icon"></span>
             <span class="dropdown-menu-text">Save search query</span>
         </button>
 
     </ux-menu>
 
+    <ux-menu #MenuItemMenu menuClass="mark-as-menu">
```
Comment (donna-marijne) [inline e2e/pages/app/menu/menu.testpage.component.html:53]:
```suggestion
    <ux-menu #markAsMenu menuClass="mark-as-menu">
```
The convention for property names is to use initial lowercase - check the documentation example.

File: e2e/tests/components/filters/filters.e2e-spec.ts [hunk 9]
```diff
@@ -235,19 +235,22 @@ describe('Filters Tests', () => {
 
         // upper case
         await page.getFilter().clear();
+        await page.clickOnAuthorMenu();
```
Comment (donna-marijne) [inline e2e/tests/components/filters/filters.e2e-spec.ts:238]: After adding the setting, these changes can be removed.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Thanks Jonathan, it works well. After using it for a bit, I think there is some risk in making this the default behavior, so let's add a `closeOnFocus` input to enable this feature.

File: e2e/tests/components/menu/menu.e2e-spec.ts [hunk 10]
```diff
@@ -42,4 +43,52 @@ describe('Menu', () => {
...
+    describe('- closeOnBlur (false)', () => {
+
+        it('should not close the menu when the tabbed away from the menu', async () => {
+            // tab to menu and open it with the keyboard
+            await page.topFocusBtn.click();
+            await browser.actions().sendKeys(Key.TAB).perform();
+            await browser.actions().sendKeys(Key.SPACE).perform();
+            await browser.actions().sendKeys(Key.TAB).perform();
+
+            // check menu closed
```
Comment (donna-marijne) [inline e2e/tests/components/menu/menu.e2e-spec.ts:55]: This comment is not correct, remove or update.

File: e2e/tests/components/menu/menu.e2e-spec.ts [hunk 11]
```diff
@@ -42,4 +43,52 @@ describe('Menu', () => {
         expect(await imageCompare('menu-dynamic-placement-left')).toEqual(0);
     });
 
-});
+    describe('- closeOnBlur (false)', () => {
+
+        it('should not close the menu when the tabbed away from the menu', async () => {
```
Comment (donna-marijne) [inline e2e/tests/components/menu/menu.e2e-spec.ts:48]:
```suggestion
        it('should not close the menu when tabbed away from the menu', async () => {
```

File: e2e/tests/components/menu/menu.e2e-spec.ts [hunk 12]
```diff
@@ -42,4 +43,52 @@ describe('Menu', () => {
...
+        });
+    });
+
+    describe('- closeOnBlur (true)', () => {
+
+        beforeEach(async () => {
+            await page.closeOnBlurBtn.click();
+        });
+
+        it('should close the menu when the tabbed away from the menu', async () => {
```
Comment (donna-marijne) [inline e2e/tests/components/menu/menu.e2e-spec.ts:66]:
```suggestion
        it('should close the menu when tabbed away from the menu', async () => {
```

**Review by donna-marijne — CHANGES_REQUESTED** [review]
The setting probably needs to be on the trigger button rather than the menu, because `closeOnBlur="true"` could be set on a submenu only and this code would not handle that (I'm not even sure what the expected behavior would be in that case).

You should also coerce boolean inputs.

**Review by donna-marijne — APPROVED** [review]
Verification:

https://plnkr.co/edit/BWsWIqbLLMjocXKC (open source) - :heavy_check_mark:
https://plnkr.co/edit/pCSS78fHQqSQFdir (Micro Focus) - :heavy_check_mark:

**Review by Roland-Penny — APPROVED** [review]
Accepted based on dev QA feedback

**Pull request merged**